### PR TITLE
Nwz/tabs for train create or edit

### LINKED ISSUE #4119: front: rearrange AddEditTrain
# AC

- All elements inside add or edit train view are ordered in 4 dedicated tabs:

1. RollingStock (more detailled ?)
2. Pathfinding
3. Allowances (in the future)
4. RunningTimeParameters :
>   - SpeedLimitByTag
>   - Ignore electrical profile
>   - PowerRestrictions

Image

### PR DESCRIPTION
**Pull request by nicolaswurtz**
Close #4119

### CODE REVIEW

File: front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx [hunk 1]
```diff
@@ -8,36 +8,34 @@ import Itinerary from 'applications/operationalStudies/components/ManageTrainSch
...
   const shouldRunPathfinding = useSelector(getShouldRunPathfinding);
   const [mustUpdatePathfinding, setMustUpdatePathfinding] = useState<boolean | undefined>(
     undefined
   );
   const { t } = useTranslation(['operationalStudies/manageTrainSchedule']);
-  const [isWorking, setIsWorking] = useState(false);
+  const rollingStockID = useSelector(getRollingStockID);
   const pathFindingID = useSelector(getPathfindingID);
-  const trainScheduleIDsToModify = useSelector(getTrainScheduleIDsToModify);
+  const trainScheduleIDsToModify: undefined | number[] = useSelector(getTrainScheduleIDsToModify);
```
Comment (clarani) [inline front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx:36]:
```suggestion
  const trainScheduleIDsToModify: number[] = useSelector(getTrainScheduleIDsToModify | []);
```

File: front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx [hunk 2]
```diff
@@ -52,40 +50,84 @@ export default function ManageTrainSchedule({ setDisplayTrainScheduleManagement
...
     }
   );
+  const tabRollingStock = {
+    title: (
+      <>
+        <img src={rollingStockPic} alt="rolling stock" height={24} />
+        <span className="ml-2">{t('tabs.rollingStock')}</span>
+      </>
+    ),
+    withWarning: rollingStockID === undefined,
```
Comment (clarani) [inline front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx:58]: (Suggestion)
```suggestion
    withWarning: !rollingStockID,
```

File: front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx [hunk 3]
```diff
@@ -52,40 +50,84 @@ export default function ManageTrainSchedule({ setDisplayTrainScheduleManagement
...
+  };
+
+  const tabPathFinding = {
+    title: (
+      <>
+        <img src={pahtFindingPic} alt="path finding" height={24} />
+        <span className="ml-2">{t('tabs.pathFinding')}</span>
+      </>
+    ),
+    withWarning: pathFindingID === undefined,
```
Comment (clarani) [inline front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx:70]: (suggestion)
```suggestion
    withWarning: !pathFindingID,
```

File: front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx [hunk 4]
```diff
@@ -52,40 +50,84 @@ export default function ManageTrainSchedule({ setDisplayTrainScheduleManagement
...
-  }
+  const tabAddingSettings = {
+    title: (
+      <>
+        <img src={addingSettingsPic} alt="adding settings" height={24} />
+        <span className="ml-2">{t('tabs.addingSettings')}</span>
+      </>
+    ),
+    label: t('tabs.addingSettings'),
+    content: !trainScheduleIDsToModify && <TrainAddingSettings />,
```
Comment (clarani) [inline front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx:127]:
```suggestion
    content: isEmpty(trainScheduleIDsToModify) && <TrainAddingSettings />,
```

File: front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx [hunk 5]
```diff
@@ -52,40 +50,84 @@ export default function ManageTrainSchedule({ setDisplayTrainScheduleManagement
...
+    title: (
+      <>
+        <img src={simulationSettings} alt="simulation settings" height={24} />
+        <span className="ml-2">{t('tabs.simulationSettings')}</span>
+      </>
+    ),
+    label: t('tabs.simulationSettings'),
+    content: (
+      <div className="row no-gutters">
+        <div className="col-lg-6 pr-lg-2">
```
Comment (clarani) [inline front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx:98]: (Suggestion)
Remove one of the 2 divs

File: front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx [hunk 6]
```diff
@@ -52,40 +50,84 @@ export default function ManageTrainSchedule({ setDisplayTrainScheduleManagement
...
+        <img src={pahtFindingPic} alt="path finding" height={24} />
+        <span className="ml-2">{t('tabs.pathFinding')}</span>
+      </>
+    ),
+    withWarning: pathFindingID === undefined,
+    label: t('tabs.pathFinding'),
+    content: (
+      <div className="row no-gutters">
+        <div className="col-xl-6 pr-xl-2">
+          {mustUpdatePathfinding !== undefined && <Itinerary mustUpdate={mustUpdatePathfinding} />}
```
Comment (clarani) [inline front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx]: Better to initialize mustUpdatePathfinding to false 
```suggestion
          <Itinerary mustUpdate={mustUpdatePathfinding} />
```

File: front/src/applications/operationalStudies/components/Scenario/TimetableManageTrainSchedule.tsx [hunk 7]
```diff
@@ -16,87 +20,68 @@ export default function TimetableManageTrainSchedule({
 }: Props) {
   const { t } = useTranslation('operationalStudies/manageTrainSchedule');
   const dispatch = useDispatch();
+  const trainScheduleIDsToModify: undefined | number[] = useSelector(getTrainScheduleIDsToModify);
```
Comment (clarani) [inline front/src/applications/operationalStudies/components/Scenario/TimetableManageTrainSchedule.tsx:23]:
```suggestion
  const trainScheduleIDsToModify: number[] = useSelector(getTrainScheduleIDsToModify | []);
```

File: front/src/common/Pathfinding/Pathfinding.tsx [hunk 8]
```diff
@@ -380,29 +380,43 @@ function Pathfinding({ mustUpdate = true, zoomToFeature }: PathfindingProps) {
...
-          {t('pathfindingError', { errorMessage: t(pathfindingState.error) })}
-        </div>
-      )}
-      {pathfindingState.missingParam && (
-        <div className="missing-params">
-          <BiErrorCircle />
-          {t('pathfindingMissingParams', { missingElements })}
-        </div>
+    <div className="pathfinding-state-main-container">
+      {Object.values(pathfindingState).every((state) => state === false || state === '') ? (
```
Comment (clarani) [inline front/src/common/Pathfinding/Pathfinding.tsx]: (suggestion)
Put this in a variable

**Review by clarani — COMMENTED** [review]
Left some comments !

File: front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx (hunk 2 above)
Comment (nicolaswurtz) [reply front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx:58]: Risk of problem if id === 0 so I'l left this as this ;)

File: front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx (hunk 3 above)
Comment (nicolaswurtz) [reply front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx:70]: idem here

File: front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx (hunk 1 above)
Comment (nicolaswurtz) [reply front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx:36]: Impossible, syntax is wrong. Moreover, trainScheduleIDsToModify is undefined when empty

File: front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx (hunk 4 above)
Comment (nicolaswurtz) [reply front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx:127]: trainScheduleIDsToModify could be undefined, so isEmpty returns an error, impossible to do :(

File: front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx (hunk 5 above)
Comment (nicolaswurtz) [reply front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx:98]: Bootstrap css conception impose use of 2 divs to respect css implementation.

File: front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx (hunk 6 above)
Comment (nicolaswurtz) [reply front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx]: Yes, was not my work but I corrected it anyway ;)

File: front/src/applications/operationalStudies/components/Scenario/TimetableManageTrainSchedule.tsx (hunk 7 above)
Comment (nicolaswurtz) [reply front/src/applications/operationalStudies/components/Scenario/TimetableManageTrainSchedule.tsx:23]: idem as above, useSelector takes function's name only

File: front/src/common/Pathfinding/Pathfinding.tsx (hunk 8 above)
Comment (nicolaswurtz) [reply front/src/common/Pathfinding/Pathfinding.tsx]: Done !

File: front/src/common/Pathfinding/Pathfinding.tsx [hunk 9]
```diff
@@ -379,9 +379,14 @@ function Pathfinding({ mustUpdate = true, zoomToFeature }: PathfindingProps) {
     [!rollingStockID, t('rollingstock')],
     [!maximumRunTime, t('maximumRunTime')],
   ]);
+
+  const isPathFindingActive = Object.values(pathfindingState).every(
```
Comment (clarani) [inline front/src/common/Pathfinding/Pathfinding.tsx:383]: Nice !! Way clearer :partying_face:

File: front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx [hunk 10]
```diff
@@ -74,7 +72,7 @@ export default function ManageTrainSchedule() {
     content: (
       <div className="row no-gutters">
         <div className="col-xl-6 pr-xl-2">
-          {mustUpdatePathfinding !== undefined && <Itinerary mustUpdate={mustUpdatePathfinding} />}
+          {mustUpdatePathfinding && <Itinerary mustUpdate={mustUpdatePathfinding} />}
```
Comment (clarani) [inline front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx]: You need to remove the test here :)
```suggestion
          <Itinerary mustUpdate={mustUpdatePathfinding} />
```

**Review by clarani — CHANGES_REQUESTED** [review]
Just one change left !

**Review by clarani — APPROVED** [review]
LGTM ! :heavy_check_mark:

**Pull request merged**